README.md

```diff
@@ -514,6 +514,86 @@ $ grpcurl -plaintext 127.0.0.1:9093 spacemesh.v1.DebugService.NetworkInfo
 }
 ```
 
+#### Handling database schema changes
+
+go-spacemesh currently maintains 2 SQLite databases in the data folder: `state.sql` (state database) and `local.sql` (local database). It employs schema versioning for both databases, with a possibility to upgrade older versions of each database to the current schema version by means of running a series of migrations. Also, go-spacemesh tracks any schema drift (unexpected schema changes) in the databases.
+
+When a database is first created, the corresponding schema file embedded in go-spacemesh executable is used to initialize it:
+* `sql/statesql/schema/schema.sql` for `state.sql`
+* `sql/localsql/schema/schema.sql` for `local.sql`
+The schema file includes `PRAGMA user_version = ...` which sets the version of the database schema. The version of the schema is equal to the number of migrations defined for the corresponding database (`state.sql` or `local.sql`).
+
+For an existing database, the `PRAGMA user_version` is checked against the expected version number. If the database's schema version is too new, go-spacemesh fails right away as an older go-spacemesh version cannot be expected to work with a database from a newer version. If the database version number is older than the expected version, go-spacemesh runs the necessary migration scripts embedded in go-spacemesh executable and updates `PRAGMA user_version = ...`. The migration scripts are located in the following folders:
+* `sql/statesql/schema/migrations` for `state.sql`
+* `sql/localsql/schema/migrations` for `local.sql`
+
+Additionally, some migrations ("coded migrations") can be implemented in Go code, in which case they reside in `.go` files located in `sql/statesql` and `sql/localsql` packages, respectively. It is worth noting that old coded migrations can be removed at some point, rendering database versions that are *too* old unusable with newer go-spacemesh versions.
```

Comment:
Putting them into `sql/statesql` and `sql/localsql` packages will almost definitly cause cyclic dependency issues. For instance any coded migration involving ATXs will most likely need to import the `activation` package, but that package already has a dependency on both`sql/statesql` and `sql/localsql` packages...

Reply:
[Last time we did it](https://github.com/spacemeshos/go-spacemesh/blob/37546bc2c93417e8dfdc24e5cc7c82d841c1f480/sql/migrations/state_0017_migration.go) we didn't have to import `activation`. In most cases `common/types` might suffice, and in the worst case a bit of code may be copied over to the migration from packages like `sql/atxs`, given that migration code has relatively short lifespan and is removed after the older versions become completely unsupported

Reply:
I seem to remember we ran into that issue with previous coded migrations for the local db.

Reply:
Fixed with a bit of extra interfaces, see here https://github.com/spacemeshos/go-spacemesh/pull/6003#issuecomment-2164193831